### ISSUE TITLE
Optional REDIS_PASSWORD env variable cause race condition and Redis NOAUTH failures

### ISSUE THREAD

**Issue by michaelvl** [argoproj/argo-helm#2836]
### Describe the bug

In PR #2800 the `REDIS_PASSWORD` environment variable was made `optional`, i.e. several `Deployment` resource have something along these lines:

```
...
        env:
          - name: REDIS_PASSWORD
            valueFrom:
              secretKeyRef:
                name: argocd-redis
                optional: true
                key: auth
```

The problem is if the secret does not exist when Pods start, then these Pods will be unable to authenticate to Redis. The Redis Pod itself does not have this env variable marked as `optional`, i,e, Redis will not startup until the secret is available.

The chart can generate the secret through a `Job` or the secret can be created by external means. Both of these can cause the secret not to be available when e.g. repo-server or application-controller Pods start.

### Related helm chart

argo-cd

### Helm chart version

Any after PR #2800, e.g. 7.3.7

### To Reproduce

Install ArgoCD with `redisSecretInit.enabled = false`, let Pods startup - e.g. repo-server and application-controller will start successfully whereas redis will restart repeatedly with 'config error' due to missing secret. Finally, create secret to let redis start. After this, one can observe 'NOAUTH Authentication required' logs in repo-server and application-controller Pods. 

### Expected behavior

Pods that must authenticate towards Redis and thus relies on the Redis secret do not start until the secret is available - like the redis Pod itself does now.

**Comment by mkilchhofer** [comment]
@shlomitubul
Could you please provide some details why you made this PR?

Also it would make sense to involve @mbevc1 and @yu-croco as you approved the implementation.
When I reviewed this PR after coming back from my vacation (after PR was merged), I thought that exactly this happens in the use case provided by issue reporter.

**Comment by yu-croco** [comment]
> as you approved the implementation.

Oh,,, I just thought the PR was trying to fix a small mistake since I didn't realize the case in this issue . 🫠

**Comment by yu-croco** [comment]
While waiting for @shlomitubul 's reply, I prepared draft PR for this issue. 🙏

**Comment by purajit** [comment]
I just came across this after upgrading to v2.11.6 and trying to figure out if I have the same issue you do.

Were you able to verify that this was indeed the issue? I checked the environment of each of the pods, and they had `REDIS_PASSWORD`. I even recycled the pods and I still have the issue. Does recycling the pods work for you?

The default behavior of `valueFrom` means that pod creation will fail if the secret is not available - that's what I observe while demo-ing too.

**Comment by purajit** [comment]
Can this be re-opened? #2839 did not fix this, as suspected.

```sh
argocd: v2.12.0+ec30a48
  BuildDate: 2024-08-05T13:46:43Z
  GitCommit: ec30a48bce7a60046836e481cd2160e28c59231d
  GitTreeState: clean
  GoVersion: go1.22.5
  Compiler: gc
  Platform: darwin/arm64
argocd-server: v2.12.0+ec30a48
  BuildDate: 2024-08-05T13:46:43Z
  GitCommit: ec30a48bce7a60046836e481cd2160e28c59231d
  GitTreeState: clean
  GoVersion: go1.22.5
  Compiler: gc
  Platform: darwin/arm64
```

**Comment by mkilchhofer** [comment]
@purajit What is your issue? Your Argo CD components cannot connect to Redis?

**Comment by purajit** [comment]
@mkilchhofer Yup, any path like `argo app diff` that relies on the cache fails with the `NOAUTH` error above; I've tried recycling the pods, I've checked their environments, and I don't see anything obviously wrong.

In particular, while basic Argo sync still works, `argocd app diff` is one functionality that depends on the cache and fails.

**Comment by mkilchhofer** [comment]
Ah we are only talking about the CLI functionality? You didn't saw any suspicious log line(s) inside the running Argo CD pods?

**Comment by purajit** [comment]
Sorry for the misunderstanding, no, since this is an issue with the overall setup, there are 
logs even internally in Argo as well. I was just giving an example of a functionality that 
obviously surfaces the issue and completely stops working - `argocd app diff` is completely 
unfunctional right now without trying to turn off Redis auth.

Here are examples of internal logs:
```	
1:S 14 Aug 2024 17:49:49.910 # MASTER aborted replication with an error: NOAUTH Authentication required.
	
1:S 14 Aug 2024 17:49:49.910 # Unexpected reply to PSYNC from master: -NOAUTH Authentication required.

1:S 14 Aug 2024 17:49:49.906 # MASTER aborted replication with an error: NOAUTH Authentication required.
	
1:S 14 Aug 2024 17:49:49.906 # Unexpected reply to PSYNC from master: -NOAUTH Authentication required.
```

**Comment by mkilchhofer** [comment]
Hi @purajit 

You shared logs of the redis-ha container or sentinel container, right?

I think the issue is related but a bit different than the one described in the initial post of this issue.

Nevertheless, could you share the values.yaml you use for your Argo CD setup, so we can try to reproduce your issue?

**Comment by purajit** [comment]
Haven't included `applicationSet`, `notifications`, `repoServer`, etc; the rest is here:
```yaml
server:
  extraArgs:
  - --insecure
  - --enable-gzip
  replicas: <...>
  resources: <...>
  serviceAccount: <...>
redis-ha:
  enabled: true
  exporter:
    enabled: false
  extraLabels:
    app.kubernetes.io/name: argocd-redis-ha-haproxy
  haproxy:
    image:
      repository: public.ecr.aws/docker/library/haproxy
  image:
    repository: public.ecr.aws/docker/library/redis
configs:
  clusterCredentials: <...>
  cm: 
    dex.config: <...>
    statusbadge.enabled: <...>
    url: <...>
  credentialTemplates: <...>
  params:
    redis.compression: none
  rbac: <...>
  repositories: <...>
  secret:
    githubSecret: <...>
```

**Comment by purajit** [comment]
https://github.com/argoproj/argo-cd/pull/19599 looks promising

**Comment by purajit** [comment]
> [argoproj/argo-cd#19599](https://github.com/argoproj/argo-cd/pull/19599)

This did indeed resolve the issue for me. It's available in 2.12.5, 2.11.10, and 2.13.0

**Comment by icdef** [comment]
I also have this problem. I use the command 
helm template argocd argo/argo-cd -f values.yaml -n argocd | kubectl apply -f -
to install argocd (since I do not want to create a helm release), so I still get some Redis NOAUTH failures

**Comment by ashinsabu3** [comment]
is there a fix here? i have seen this sometimes happen also if the job doesnt complete before the pods start for repo-server abd controller and its intermittent and we end up having to manually bounce the argocd workloads

**Comment by andreas-p** [comment]
I also encounter NOAUTH problems after upgrading from 7.7.0 (Argo-CD 2.13.0) to 7.8.2 (Argo-CD 2.14.2).
Narrowing my problem further, I found that my deployment working stops with 7.7.6, which includes the optional flag from #3058. 

After patching 7.8.2 to have `optional: false` on three REDIS_PASSWORD occurrences, my deployment works again.

So either #3058 needs reverted, or the optional flag optionally configurable.

**Comment by Makeshift** [comment]
This also seems to affect any workflow that doesn't utilize Helm hooks, eg in my case deploying with Pulumi. The Pulumi v4 chart [completely disregards](https://github.com/pulumi/pulumi-kubernetes/issues/3284) any resources with Helm hooks, meaning the secret is never created and any pods requiring it crash loop.

**Comment by ashinsabu3** [comment]
perhaps this can be fixed by adding this check for optional to not be true.

1. secretInit enabled true in non-ha redis  https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/values.yaml#L1681 

OR 

2. redis-ha enabled https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/values.yaml#L1574 
    AND 
    redis-ha auth enabled https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/values.yaml#L1627

@mkilchhofer

**Comment by mkilchhofer** [comment]
Created another PR here:
- #3228

Maybe we can get some feedback about it. I know some code is now less optimized but it should be more clear now in what cases we require the existence of the Secret.
